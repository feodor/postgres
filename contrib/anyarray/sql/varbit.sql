SELECT
    t,
	ARRAY(
		SELECT
			v::bit(10)::varbit
		FROM
			generate_series(max(0, t - 10),  t) as v
	) AS v
	INTO test_varbit
FROM
	generate_series(1, 200) as t;


SET anyarray.similarity_type=cosine;
SELECT  t, similarity(v, '{10,9,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[]) AS s FROM test_varbit 
	WHERE v % '{10,9,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[] ORDER BY s DESC, t;
SELECT  t, similarity(v, '{50,49,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[]) AS s FROM test_varbit 
	WHERE v % '{50,49,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[] ORDER BY s DESC, t;

SET anyarray.similarity_type=jaccard;
SELECT  t, similarity(v, '{10,9,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[]) AS s FROM test_varbit 
	WHERE v % '{10,9,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[] ORDER BY s DESC, t;
SELECT  t, similarity(v, '{50,49,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[]) AS s FROM test_varbit 
	WHERE v % '{50,49,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[] ORDER BY s DESC, t;

SELECT t, v FROM test_varbit WHERE v && '{43,50}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
SELECT t, v FROM test_varbit WHERE v @> '{43,50}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
SELECT t, v FROM test_varbit WHERE v <@ '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
SELECT t, v FROM test_varbit WHERE v =  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
SET anyarray.similarity_type=cosine;
SELECT t, v FROM test_varbit WHERE v %  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
SET anyarray.similarity_type=jaccard;
SELECT t, v FROM test_varbit WHERE v %  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
SET anyarray.similarity_type=overlap;
SET anyarray.similarity_threshold = 3;
SELECT t, v FROM test_varbit WHERE v %  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
RESET anyarray.similarity_threshold;

CREATE INDEX idx_test_varbit ON test_varbit USING gin (v _varbit_aa_ops);

SET enable_seqscan=off;

EXPLAIN (COSTS OFF) SELECT t, v FROM test_varbit WHERE v && '{43,50}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
EXPLAIN (COSTS OFF) SELECT t, v FROM test_varbit WHERE v @> '{43,50}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
EXPLAIN (COSTS OFF) SELECT t, v FROM test_varbit WHERE v <@ '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
EXPLAIN (COSTS OFF) SELECT t, v FROM test_varbit WHERE v =  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
EXPLAIN (COSTS OFF) SELECT t, v FROM test_varbit WHERE v %  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;

SELECT t, v FROM test_varbit WHERE v && '{43,50}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
SELECT t, v FROM test_varbit WHERE v @> '{43,50}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
SELECT t, v FROM test_varbit WHERE v <@ '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
SELECT t, v FROM test_varbit WHERE v =  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
SET anyarray.similarity_type=cosine;
SELECT t, v FROM test_varbit WHERE v %  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
SET anyarray.similarity_type=jaccard;
SELECT t, v FROM test_varbit WHERE v %  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
SET anyarray.similarity_type=overlap;
SET anyarray.similarity_threshold = 3;
SELECT t, v FROM test_varbit WHERE v %  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
RESET anyarray.similarity_threshold;

