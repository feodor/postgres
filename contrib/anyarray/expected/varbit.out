SELECT
    t,
	ARRAY(
		SELECT
			v::bit(10)::varbit
		FROM
			generate_series(max(0, t - 10),  t) as v
	) AS v
	INTO test_varbit
FROM
	generate_series(1, 200) as t;
SET anyarray.similarity_type=cosine;
SELECT  t, similarity(v, '{10,9,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[]) AS s FROM test_varbit 
	WHERE v % '{10,9,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[] ORDER BY s DESC, t;
 t  |    s     
----+----------
 10 | 0.953463
 11 | 0.953463
  9 |      0.9
 12 | 0.858116
  8 | 0.843274
  7 | 0.782624
 13 |  0.76277
  6 | 0.717137
 14 | 0.667424
  5 | 0.645497
(10 rows)

SELECT  t, similarity(v, '{50,49,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[]) AS s FROM test_varbit 
	WHERE v % '{50,49,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[] ORDER BY s DESC, t;
 t  |    s     
----+----------
  8 | 0.843274
  9 |      0.8
  7 | 0.782624
 10 |  0.76277
 11 |  0.76277
  6 | 0.717137
 12 | 0.667424
  5 | 0.645497
(8 rows)

SET anyarray.similarity_type=jaccard;
SELECT  t, similarity(v, '{10,9,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[]) AS s FROM test_varbit 
	WHERE v % '{10,9,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[] ORDER BY s DESC, t;
 t  |    s     
----+----------
 10 | 0.909091
 11 | 0.909091
  9 | 0.818182
 12 |     0.75
  8 | 0.727273
  7 | 0.636364
 13 | 0.615385
(7 rows)

SELECT  t, similarity(v, '{50,49,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[]) AS s FROM test_varbit 
	WHERE v % '{50,49,8,7,6,5,4,3,2,1}'::int4[]::bit(10)[]::varbit[] ORDER BY s DESC, t;
 t  |    s     
----+----------
  8 | 0.727273
  9 | 0.666667
  7 | 0.636364
 10 | 0.615385
 11 | 0.615385
(5 rows)

SELECT t, v FROM test_varbit WHERE v && '{43,50}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
 t  |                                                             v                                                              
----+----------------------------------------------------------------------------------------------------------------------------
 43 | {0000100001,0000100010,0000100011,0000100100,0000100101,0000100110,0000100111,0000101000,0000101001,0000101010,0000101011}
 44 | {0000100010,0000100011,0000100100,0000100101,0000100110,0000100111,0000101000,0000101001,0000101010,0000101011,0000101100}
 45 | {0000100011,0000100100,0000100101,0000100110,0000100111,0000101000,0000101001,0000101010,0000101011,0000101100,0000101101}
 46 | {0000100100,0000100101,0000100110,0000100111,0000101000,0000101001,0000101010,0000101011,0000101100,0000101101,0000101110}
 47 | {0000100101,0000100110,0000100111,0000101000,0000101001,0000101010,0000101011,0000101100,0000101101,0000101110,0000101111}
 48 | {0000100110,0000100111,0000101000,0000101001,0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000}
 49 | {0000100111,0000101000,0000101001,0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001}
 50 | {0000101000,0000101001,0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010}
 51 | {0000101001,0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011}
 52 | {0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011,0000110100}
 53 | {0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011,0000110100,0000110101}
 54 | {0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011,0000110100,0000110101,0000110110}
 55 | {0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011,0000110100,0000110101,0000110110,0000110111}
 56 | {0000101110,0000101111,0000110000,0000110001,0000110010,0000110011,0000110100,0000110101,0000110110,0000110111,0000111000}
 57 | {0000101111,0000110000,0000110001,0000110010,0000110011,0000110100,0000110101,0000110110,0000110111,0000111000,0000111001}
 58 | {0000110000,0000110001,0000110010,0000110011,0000110100,0000110101,0000110110,0000110111,0000111000,0000111001,0000111010}
 59 | {0000110001,0000110010,0000110011,0000110100,0000110101,0000110110,0000110111,0000111000,0000111001,0000111010,0000111011}
 60 | {0000110010,0000110011,0000110100,0000110101,0000110110,0000110111,0000111000,0000111001,0000111010,0000111011,0000111100}
(18 rows)

SELECT t, v FROM test_varbit WHERE v @> '{43,50}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
 t  |                                                             v                                                              
----+----------------------------------------------------------------------------------------------------------------------------
 50 | {0000101000,0000101001,0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010}
 51 | {0000101001,0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011}
 52 | {0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011,0000110100}
 53 | {0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011,0000110100,0000110101}
(4 rows)

SELECT t, v FROM test_varbit WHERE v <@ '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
 t  |                                                             v                                                              
----+----------------------------------------------------------------------------------------------------------------------------
  1 | {0000000000,0000000001}
  2 | {0000000000,0000000001,0000000010}
  3 | {0000000000,0000000001,0000000010,0000000011}
  4 | {0000000000,0000000001,0000000010,0000000011,0000000100}
  5 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101}
  6 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110}
  7 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111}
  8 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000}
  9 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001}
 10 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}
(10 rows)

SELECT t, v FROM test_varbit WHERE v =  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
 t  |                                                             v                                                              
----+----------------------------------------------------------------------------------------------------------------------------
 10 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}
(1 row)

SET anyarray.similarity_type=cosine;
SELECT t, v FROM test_varbit WHERE v %  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
 t  |                                                             v                                                              
----+----------------------------------------------------------------------------------------------------------------------------
  3 | {0000000000,0000000001,0000000010,0000000011}
  4 | {0000000000,0000000001,0000000010,0000000011,0000000100}
  5 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101}
  6 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110}
  7 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111}
  8 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000}
  9 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001}
 10 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}
 11 | {0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011}
 12 | {0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100}
 13 | {0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100,0000001101}
 14 | {0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100,0000001101,0000001110}
(12 rows)

SET anyarray.similarity_type=jaccard;
SELECT t, v FROM test_varbit WHERE v %  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
 t  |                                                             v                                                              
----+----------------------------------------------------------------------------------------------------------------------------
  6 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110}
  7 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111}
  8 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000}
  9 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001}
 10 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}
 11 | {0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011}
 12 | {0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100}
(7 rows)

SET anyarray.similarity_type=overlap;
SET anyarray.similarity_threshold = 3;
SELECT t, v FROM test_varbit WHERE v %  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
 t  |                                                             v                                                              
----+----------------------------------------------------------------------------------------------------------------------------
  2 | {0000000000,0000000001,0000000010}
  3 | {0000000000,0000000001,0000000010,0000000011}
  4 | {0000000000,0000000001,0000000010,0000000011,0000000100}
  5 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101}
  6 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110}
  7 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111}
  8 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000}
  9 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001}
 10 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}
 11 | {0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011}
 12 | {0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100}
 13 | {0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100,0000001101}
 14 | {0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100,0000001101,0000001110}
 15 | {0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100,0000001101,0000001110,0000001111}
 16 | {0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100,0000001101,0000001110,0000001111,0000010000}
 17 | {0000000111,0000001000,0000001001,0000001010,0000001011,0000001100,0000001101,0000001110,0000001111,0000010000,0000010001}
 18 | {0000001000,0000001001,0000001010,0000001011,0000001100,0000001101,0000001110,0000001111,0000010000,0000010001,0000010010}
(17 rows)

RESET anyarray.similarity_threshold;
CREATE INDEX idx_test_varbit ON test_varbit USING gin (v _varbit_aa_ops);
SET enable_seqscan=off;
EXPLAIN (COSTS OFF) SELECT t, v FROM test_varbit WHERE v && '{43,50}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
                                QUERY PLAN                                 
---------------------------------------------------------------------------
 Sort
   Sort Key: t
   ->  Bitmap Heap Scan on test_varbit
         Recheck Cond: (v && '{0000101011,0000110010}'::bit varying[])
         ->  Bitmap Index Scan on idx_test_varbit
               Index Cond: (v && '{0000101011,0000110010}'::bit varying[])
(6 rows)

EXPLAIN (COSTS OFF) SELECT t, v FROM test_varbit WHERE v @> '{43,50}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
                                QUERY PLAN                                 
---------------------------------------------------------------------------
 Sort
   Sort Key: t
   ->  Bitmap Heap Scan on test_varbit
         Recheck Cond: (v @> '{0000101011,0000110010}'::bit varying[])
         ->  Bitmap Index Scan on idx_test_varbit
               Index Cond: (v @> '{0000101011,0000110010}'::bit varying[])
(6 rows)

EXPLAIN (COSTS OFF) SELECT t, v FROM test_varbit WHERE v <@ '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
                                                                                  QUERY PLAN                                                                                  
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort
   Sort Key: t
   ->  Bitmap Heap Scan on test_varbit
         Recheck Cond: (v <@ '{0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}'::bit varying[])
         ->  Bitmap Index Scan on idx_test_varbit
               Index Cond: (v <@ '{0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}'::bit varying[])
(6 rows)

EXPLAIN (COSTS OFF) SELECT t, v FROM test_varbit WHERE v =  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
                                                                                 QUERY PLAN                                                                                  
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort
   Sort Key: t
   ->  Bitmap Heap Scan on test_varbit
         Recheck Cond: (v = '{0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}'::bit varying[])
         ->  Bitmap Index Scan on idx_test_varbit
               Index Cond: (v = '{0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}'::bit varying[])
(6 rows)

EXPLAIN (COSTS OFF) SELECT t, v FROM test_varbit WHERE v %  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
                                                                                 QUERY PLAN                                                                                  
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort
   Sort Key: t
   ->  Bitmap Heap Scan on test_varbit
         Recheck Cond: (v % '{0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}'::bit varying[])
         ->  Bitmap Index Scan on idx_test_varbit
               Index Cond: (v % '{0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}'::bit varying[])
(6 rows)

SELECT t, v FROM test_varbit WHERE v && '{43,50}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
 t  |                                                             v                                                              
----+----------------------------------------------------------------------------------------------------------------------------
 43 | {0000100001,0000100010,0000100011,0000100100,0000100101,0000100110,0000100111,0000101000,0000101001,0000101010,0000101011}
 44 | {0000100010,0000100011,0000100100,0000100101,0000100110,0000100111,0000101000,0000101001,0000101010,0000101011,0000101100}
 45 | {0000100011,0000100100,0000100101,0000100110,0000100111,0000101000,0000101001,0000101010,0000101011,0000101100,0000101101}
 46 | {0000100100,0000100101,0000100110,0000100111,0000101000,0000101001,0000101010,0000101011,0000101100,0000101101,0000101110}
 47 | {0000100101,0000100110,0000100111,0000101000,0000101001,0000101010,0000101011,0000101100,0000101101,0000101110,0000101111}
 48 | {0000100110,0000100111,0000101000,0000101001,0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000}
 49 | {0000100111,0000101000,0000101001,0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001}
 50 | {0000101000,0000101001,0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010}
 51 | {0000101001,0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011}
 52 | {0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011,0000110100}
 53 | {0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011,0000110100,0000110101}
 54 | {0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011,0000110100,0000110101,0000110110}
 55 | {0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011,0000110100,0000110101,0000110110,0000110111}
 56 | {0000101110,0000101111,0000110000,0000110001,0000110010,0000110011,0000110100,0000110101,0000110110,0000110111,0000111000}
 57 | {0000101111,0000110000,0000110001,0000110010,0000110011,0000110100,0000110101,0000110110,0000110111,0000111000,0000111001}
 58 | {0000110000,0000110001,0000110010,0000110011,0000110100,0000110101,0000110110,0000110111,0000111000,0000111001,0000111010}
 59 | {0000110001,0000110010,0000110011,0000110100,0000110101,0000110110,0000110111,0000111000,0000111001,0000111010,0000111011}
 60 | {0000110010,0000110011,0000110100,0000110101,0000110110,0000110111,0000111000,0000111001,0000111010,0000111011,0000111100}
(18 rows)

SELECT t, v FROM test_varbit WHERE v @> '{43,50}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
 t  |                                                             v                                                              
----+----------------------------------------------------------------------------------------------------------------------------
 50 | {0000101000,0000101001,0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010}
 51 | {0000101001,0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011}
 52 | {0000101010,0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011,0000110100}
 53 | {0000101011,0000101100,0000101101,0000101110,0000101111,0000110000,0000110001,0000110010,0000110011,0000110100,0000110101}
(4 rows)

SELECT t, v FROM test_varbit WHERE v <@ '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
 t  |                                                             v                                                              
----+----------------------------------------------------------------------------------------------------------------------------
  1 | {0000000000,0000000001}
  2 | {0000000000,0000000001,0000000010}
  3 | {0000000000,0000000001,0000000010,0000000011}
  4 | {0000000000,0000000001,0000000010,0000000011,0000000100}
  5 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101}
  6 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110}
  7 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111}
  8 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000}
  9 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001}
 10 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}
(10 rows)

SELECT t, v FROM test_varbit WHERE v =  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
 t  |                                                             v                                                              
----+----------------------------------------------------------------------------------------------------------------------------
 10 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}
(1 row)

SET anyarray.similarity_type=cosine;
SELECT t, v FROM test_varbit WHERE v %  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
 t  |                                                             v                                                              
----+----------------------------------------------------------------------------------------------------------------------------
  3 | {0000000000,0000000001,0000000010,0000000011}
  4 | {0000000000,0000000001,0000000010,0000000011,0000000100}
  5 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101}
  6 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110}
  7 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111}
  8 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000}
  9 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001}
 10 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}
 11 | {0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011}
 12 | {0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100}
 13 | {0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100,0000001101}
 14 | {0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100,0000001101,0000001110}
(12 rows)

SET anyarray.similarity_type=jaccard;
SELECT t, v FROM test_varbit WHERE v %  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
 t  |                                                             v                                                              
----+----------------------------------------------------------------------------------------------------------------------------
  6 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110}
  7 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111}
  8 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000}
  9 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001}
 10 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}
 11 | {0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011}
 12 | {0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100}
(7 rows)

SET anyarray.similarity_type=overlap;
SET anyarray.similarity_threshold = 3;
SELECT t, v FROM test_varbit WHERE v %  '{0,1,2,3,4,5,6,7,8,9,10}'::int4[]::bit(10)[]::varbit[] ORDER BY t;
 t  |                                                             v                                                              
----+----------------------------------------------------------------------------------------------------------------------------
  2 | {0000000000,0000000001,0000000010}
  3 | {0000000000,0000000001,0000000010,0000000011}
  4 | {0000000000,0000000001,0000000010,0000000011,0000000100}
  5 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101}
  6 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110}
  7 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111}
  8 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000}
  9 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001}
 10 | {0000000000,0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010}
 11 | {0000000001,0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011}
 12 | {0000000010,0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100}
 13 | {0000000011,0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100,0000001101}
 14 | {0000000100,0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100,0000001101,0000001110}
 15 | {0000000101,0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100,0000001101,0000001110,0000001111}
 16 | {0000000110,0000000111,0000001000,0000001001,0000001010,0000001011,0000001100,0000001101,0000001110,0000001111,0000010000}
 17 | {0000000111,0000001000,0000001001,0000001010,0000001011,0000001100,0000001101,0000001110,0000001111,0000010000,0000010001}
 18 | {0000001000,0000001001,0000001010,0000001011,0000001100,0000001101,0000001110,0000001111,0000010000,0000010001,0000010010}
(17 rows)

RESET anyarray.similarity_threshold;
